x/concentrated-liquidity/swaps.go

```diff
@@ -749,3 +757,22 @@ func (k Keeper) getSwapAccumulators(ctx sdk.Context, poolId uint64) (accum.Accum
 	}
 	return spreadAccum, uptimeAccums, nil
 }
+
+// validateSwapProgressAndAmountConsumption validates that the swap progress and amount consumption are valid. These are valid if:
+// - computedSqrtPrice is not equal to sqrtPriceStart (progress made)
+// - computedSqrtPrice is equals to sqrtPriceStart and both amountIn and amountOut are zero (progress not made AND amounts are not consumed)
+// If swap succeeded within the same ULP while consuming amounts, that would mean that a trader can exploit the pool to
+// continue getting the same execution price. To prevent that, we add this check that fails the swap if amounts
+// in or out are non-zero while swap does not move even by 1 ULP.
+// If no progress is made with zero amounts, there is a risk of running into an infinite loop which we prevent
+// by a constant bound on the the number of iterations without progress.
+// Note, that it is correct to have a single iteration without progress and zero amounts. One valid example
+// is swapping one for zero (right) directly to the tick. Then, immediately swapping zero for one (left).
+// After the fist swap, out sqrtPriceCurrent is the crossed tick's sqrt price. Also sqrtPriceTarget is the crossed tick's sqrt price.
+// In such a case, no amounts are consumed and the swap step is allowed to succeed.
+func validateSwapProgressAndAmountConsumption(computedSqrtPrice, sqrtPriceStart osmomath.BigDec, amountIn, amountOut sdk.Dec) error {
+	if computedSqrtPrice.Equal(sqrtPriceStart) && !(amountIn.IsZero() && amountOut.IsZero()) {
```

Comment:
What was the reasoning behind adding the second conditional on nonzero amounts here?

Reply:
There are valid instances where it is valid to have both amounts as zero.

```
Note that having a single iteration without progress and zero amounts is correct. One good example
is swapping one for zero (right) directly to the tick. Then, immediately swapping zero for one (left).
After the first swap, our sqrtPriceCurrent is the crossed tick's sqrt price. Also, sqrtPriceTarget is the crossed tick's sqrt price. In such a case, no amounts are consumed and the swap step is allowed to succeed.
```